Quaver.Shared/Screens/Edit/Actions/HitObjects/Resnap/EditorActionResnapHitObjects.cs

```diff
@@ -0,0 +1,128 @@
+            foreach (var note in WorkingMap.HitObjects)
+            {
+                var startTimeDelta = DiffToClosestSnap(note.StartTime);
+                var endTimeDelta = DiffToClosestSnap(note.EndTime);
+                var startTimeNotSnapped = Math.Abs(startTimeDelta) >= 1;
+                var endTimeNotSnapped = Math.Abs(endTimeDelta) >= 1 && note.IsLongNote;
+
+                if (!startTimeNotSnapped && !endTimeNotSnapped)
+                    continue;
+
+                var resnappedNote = Helpers.ObjectHelper.DeepClone(note);
+
+                if (startTimeNotSnapped)
+                {
+                    resnappedNote.StartTime = (int)(resnappedNote.StartTime - startTimeDelta);
+                    Logger.Debug($"Resnapped {note.StartTime}|{note.Lane} to {resnappedNote.StartTime}", LogType.Runtime, false);
+                }
+
+                if (endTimeNotSnapped)
+                {
+                    resnappedNote.EndTime = (int)(resnappedNote.EndTime - endTimeDelta);
+                    Logger.Debug($"Resnapped {note.EndTime}|{note.Lane} to {resnappedNote.EndTime} (Long Note End)", LogType.Runtime, false);
+                }
+
+                OldNotes.Add(note);
+                NewNotes.Add(resnappedNote);
+            }
+
+            if (OldNotes.Count() > 0)
+            {
+                OldNotes.ForEach(n => WorkingMap.HitObjects.Remove(n));
+                WorkingMap.HitObjects.AddRange(NewNotes);
+                WorkingMap.Sort();
+                NotificationManager.Show(NotificationLevel.Info, $"Resnapped {NewNotes.Count} notes");
+                ActionManager.TriggerEvent(EditorActionType.ResnapHitObjects, new EditorActionHitObjectsResnappedEventArgs(Snaps));
+            }
+            else
+                NotificationManager.Show(NotificationLevel.Info, $"No notes to resnap");
+
+        }
+
+        private float DiffToClosestSnap(int time) {
+            var timingPoint = WorkingMap.GetTimingPointAt(time);
+            var msPerSnaps = Snaps.Select(s => timingPoint.MillisecondsPerBeat / s).ToList();
+
+            var smallestDelta = float.MaxValue;
+            foreach (var msPerSnap in msPerSnaps)
+            {
+                var deltaForward = (time - timingPoint.StartTime) % msPerSnap;
+                var deltaBackward = deltaForward - msPerSnap;
+                var delta = deltaForward < -deltaBackward ? deltaForward : deltaBackward;
+                if (Math.Abs(delta) < Math.Abs(smallestDelta))
+                    smallestDelta = delta;
+            }
+
+            return smallestDelta;
+        }
+
+        /// <inheritdoc />
+        /// <summary>
+        /// </summary>
+        public void Undo()
+        {
+            NewNotes.ForEach(n => WorkingMap.HitObjects.Remove(n));
+            WorkingMap.HitObjects.AddRange(OldNotes);
+            WorkingMap.Sort();
+        }
```

Comment:
Few issues with this:

1. Undoing and redoing adds the same notes to the list and more objects to the map - ![img](https://i.imgur.com/q5ZKN40.png)
2. The event for this action isn't handled in `EditorPlayfeld`, so notes that are moved by this do not update.
3. The event for this action isn't handled in `EditorMapPreview`, so performing this action will not refresh the gameplay preview.
4. The event for this action isn't handled in `EditorPanelDetails`, so the object count & difficulty rating are never updated.
5. The event isn't nullified on `Dispose()` in `EditorActionManager` - causing a memory leak.